ebpf: delete existing pinned map if incompatible with the spec

jibi opened the pull request:
This PR updates the OpenOrCreate() method of the ebpf package to
gracefully handle the case of a pinned map that is incompatible with the
map's spec passed to the method.

Rather than just returning the error returned by NewMapWithOptions(), we
now log a warning and try to delete and recreate the pinned map.

This mimics the behaviour of the old OpenOrCreate() method of the bpf
package.

Fixes: #15629

jibi committed:
go.mod: bump cilium/ebpf

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

aanm commented on pkg/ebpf/map.go:
```diff
@@ -70,13 +71,42 @@ func (m *Map) OpenOrCreate() error {
+			return fmt.Errorf("unable to create map: %w", err)
+		}
+
+		// There already exists a pinned map but it has a different
+		// configuration (e.g different type, k/v size or flags).
+		// Try to delete and recreate it.
```
Is it safe to do this for all maps? I also assume this can't be done "atomically"?

jibi replied:
> Is it safe to do this for all maps?

That's what we are already doing with all the other maps that are opened using the `bpf.OpenOrCreate()` method:

https://github.com/cilium/cilium/blob/648f4f05c46abb98d3f31bd5610db283d0cd3953/pkg/bpf/map_linux.go#L503-L505

~I'm not saying that this is necessarily the best approach, but since the aim of the `ebpf` package is to (at least initially) mimic the `bpf` package (to avoid introducing another dimension of complexity while we migrate to that) I think we should stick with this approach for now~

edit: I just found that the `bpf` package has also an `Open()` method (in addition to the `OpenOrCreate()`) which will return an error in case the map's spec is incompatible with the existing map

> I also assume this can't be done "atomically"?

That's my understanding, yes

christarazi reviewed (commented):
I haven't looked deeply at the code yet, but I glanced quickly and couldn't tell if the PR addresses this comment: https://github.com/cilium/cilium/issues/15629#issuecomment-816942916

christarazi commented on pkg/ebpf/map.go:
```diff
@@ -70,13 +71,42 @@ func (m *Map) OpenOrCreate() error {
+
+		oldMap, err := ciliumebpf.LoadPinnedMap(path, &opts.LoadPinOptions)
+		if err != nil {
+			return fmt.Errorf("Cannot load pinned map %s: %w", m.spec.Name, err)
```
Typically error strings should be lower-cased: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

christarazi reviewed (changes requested):
LGTM, one minor nit

joestringer commented on pkg/ebpf/map.go:
```diff
@@ -70,13 +71,42 @@ func (m *Map) OpenOrCreate() error {
+		// Try to delete and recreate it.
+
+		log.WithField("map", m.spec.Name).
+			Warn("cannot load pinned map as it's incompatible with the spec provided. Deleting and recreating it.")
```
A few comments here:

* The current implementation, via `objCheck()`, logs specifically which attributes are incompatible. This means that we have some chance of understanding what the incompatibility is and this can go a long way to fixing unintended breakages introduced by developers. Without that, since this code is deleting the old map, it can be quite difficult to understand exactly what is going on because the old map is gone by the time you go to inspect it.
* The current equivalent log message uses the wording `Removing map to allow for property upgrade (expect map data loss)`, basically to try to raise the alarm that this is really not expected. We can maybe argue if it's too alarming given that I believe only CT map deletion would cause actual packet impact, but I wanted to at least check whether there was any discussion about this? Even if we used the exact same wording, we should be able to differentiate log messages from the old package vs. the new package by checking the `subsys` field (should be `ebpf` for the new version, or `bpf` for the old one).
* nit, while errors should start with lower case (since they are often concatenated with other messages), log messages usually start with a capital case.

jibi replied:
I updated the message to include the error from `cilium/ebpf` that indicates which attribute is incompatible with the spec and reintroduced the wording of the previous log. The user should now get something like:

```
level=warning msg="Removing map to allow for property upgrade (expect map data loss): expected flags 1, got 0: map's spec is incompatible with pinned map" map=cilium_metrics subsys=bpf
```

joestringer commented on pkg/ebpf/map.go:
```diff
@@ -70,13 +71,42 @@ func (m *Map) OpenOrCreate() error {
+
+		if err = oldMap.Close(); err != nil {
+			return fmt.Errorf("cannot close map %s: %w", m.spec.Name, err)
+		}
```
The error handling seems a bit unusual here: As soon as `LoadPinnedMap()` has returned without error, the fd is open. If the `Unpin()` theoretically failed then we wouldn't close the map. Should we just use `defer` here immediately after the successful LoadPinnedMap?

Also not sure how `oldMap.Close()` can fail but if we were able to `Unpin()` the old one, do we really need to abort the creation of the new map in that error condition? Maybe we can just log a warning and continue? (idea being if there is some really weird behaviour here then at least it will show up in the logs, but it shouldn't prevent the broader `OpenOrCreate()` operation that's being requested)

joestringer commented on pkg/ebpf/map.go:
```diff
@@ -70,13 +71,42 @@ func (m *Map) OpenOrCreate() error {
 	mapType := bpf.GetMapType(bpf.MapType(m.spec.Type))
 	m.spec.Flags = m.spec.Flags | bpf.GetPreAllocateMapFlags(mapType)
 
+	path := bpf.MapPath(m.spec.Name)
```
I think we're still missing one other difference between the implementations: `pkg/bpf/bpf_linux.go:OpenOrCreateMap()` will attempt to create the parent directory here before pinning; from a brief search through the ebpf library I didn't see the same logic.

Might not be too relevant for this PR depending on the order of map creation but we'll want to fix this up at some point before we fully switch libraries over.

jibi replied:
Yep, I initially ignored it as `cilium_metrics` is not the first pinned map opened by the agent, but since we are on it I just added the logic to `ebpf.OpenOrCreate()`

joestringer reviewed (changes requested):
Thanks for the recap, that was useful framing to consider these changes.

I went through the call stack for both cases and noticed a couple of remaining minor differences which I've highlighted below.

jibi committed:
ebpf: delete existing pinned map if incompatible with the spec

This commit updates the OpenOrCreate() method of the ebpf package to
gracefully handle the case of a pinned map that is incompatible with the
map's spec passed to the method.

Rather than just returning the error returned by NewMapWithOptions(), we
now log a warning and try to delete and recreate the pinned map.

This mimics the behaviour of the old OpenOrCreate() method of the bpf
package.

Fixes: #15629

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

jibi committed:
ebpf: use "ebpf" as subsys name for logs

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

jibi committed:
ebpf: create map's parent directory if it doesn't exist

When dealing with a pinned map in the ebpf.OpenOrCreate method, make
sure the map's parent directory does exist (and try to create it in case
it doesn't).

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

joestringer commented on pkg/ebpf/map.go:
```diff
@@ -70,13 +73,57 @@ func (m *Map) OpenOrCreate() error {
+		// Try to delete and recreate it.
+
+		log.WithField("map", m.spec.Name).
+			Warnf("Removing map to allow for property upgrade (expect map data loss): %v", errors.Unwrap(err))
```
Usually we use `.WithError()` to include an individual error, so that if a log shows up multiple times with different parameters, it is possible to correlate the common pieces in structured logging systems (eg the message is the same across all, but the error itself may vary). This can also help with grepping for an error as we can always tell that the `msg=` part is static text:

```suggestion
		log.WithField("map", m.spec.Name).
			WithError(err).Warn("Removing map to allow for property upgrade (expect map data loss)")
```

ref. https://github.com/cilium/cilium/issues/14156

joestringer reviewed (approved):
Thanks, just one remaining nit (with a few occurrences; we used to lint this out of the codebase but it seems we lost that linter somewhere along the way). Once that's addressed, LGTM.
